Proposal: GraphQL Composite Schemas Working Group

benjie opened the pull request:
There's a _lot_ of different ways of combining multiple GraphQL schemas together to form a larger composite schema, and the wide variety of options can make it challenging to design schemas in a way that will be easy to compose later. To name but a few approaches:

* Schema merging and stitching (various implementations in many languages)
* GraphQL modules and other "extend type"-based approaches
* Federation (Apollo's v1 and v2, Mercurius, WunderGraph, Hot Chocolate, etc)
* Hasura's GraphQL Joins

Though these are all separate solutions to similar problems, there are various concerns that most of them have to consider:

* Identifying the schemas to join
* Discovering/defining the relationships between the types/schemas
* Schema manipulation to avoid conflicts (e.g. renaming types/fields)
* Declaring which schemas "own" a particular type, and detecting conflicts
* Detecting breaking changes
* Actually fetching the relevant data, and combining it to fulfil the GraphQL request
* etc

Not wanting to cause an [XKCD#927](https://xkcd.com/927/) situation…

![XKCD 927](https://user-images.githubusercontent.com/129910/166689080-bc45979a-6e70-4d4f-9bcc-f8f3daea1ee6.png)

… but, it feels like it would be of benefit to the ecosystem at large if there were a shared specification that covers a few of these needs.

---

I propose we set up a working group encompassing people working on these problems and start some discussions around forming a common specification that covers a few of these needs.

I'm not suggesting that we aim to write a full specification for "The One True Way To Build A Big GraphQL Schema From A Number Of Smaller GraphQL Schemas" (TOTWTBABGSFANOSGS) ─ the different approaches listed above tend to solve slightly different problems and have different pros and cons for different situations. Instead, I propose a smaller specification that describes the common ground that everyone can build and innovate on top of. My hope is that this will help the entire ecosystem iterate more rapidly; and will also give GraphQL users more choice, enabling the best solutions for each given problem set to float to the surface more easily.

Initially, I think it would be great to get representatives from the various projects working in this space together to discuss what a spec might look like, and what it should contain. I propose the name "GraphQL Composite Schemas Working Group" to keep it as vendor neutral as possible.

If you're interested, comment below!

yaacovCR commented:
I would love to join the group.

rivantsov commented:
add me too, thanks!

timsuchanek commented:
Thanks for kicking this off @benjie! I'd love to join!

saihaj commented:
Yes I would love to join this!

hayes commented:
I'm interested in working on this

mcollina commented:
I’m in

jensneuse commented:
CEO of [WunderGraph](https://wundergraph.com) here. I really like the idea and would love to participate/contribute!
If this spec look right, we'd be more than happy to contribute a MIT licensed Gateway implementation in go/golang and add it to [graphql-go-tools](https://github.com/wundergraph/graphql-go-tools).

Urigo commented:
So happy this is happening, I would love to join and support, thanks for leading it @benjie !

xuorig commented:
Count me in 👍

grillorafael commented:
Hey, I know I am not active here or anything but I've been using GraphQL in an enterprise context for about 2 years and relying on Federation / Stitch for a while. Have a special interest on Subscriptions and how people develop distributed schemas independently as well.

Any way I can join and be part of this working group?

benjie commented:
Working groups tend to be open to all; this WG doesn't exist yet, but if/when it does I'll post here with details :+1: 

(Also thanks everyone for chiming in! I'm raising this at the next GraphQL Spec WG to see if we can make it an official WG :+1: )

obi1kenobi commented:
Great idea, would love to join! 🙌

praveenweb commented:
@benjie - Great initiative! Would love to participate and contribute to an open spec :)

patrick91 commented:
I’m interested too :)

ndejaco2 commented:
I am interested :)

notrab commented:
This is awesome. I'd love to be part of it, and help in any way I can to create some content on this once we have something 😄

charlypoly commented:
Interested too, great initiative!

yaacovCR commented on rfcs/CompositeSchemas.md:
```diff
@@ -0,0 +1,26 @@
+* Identifying the schemas to join
+* Discovering/defining the relationships between the types/schemas
+* Schema manipulation to avoid conflicts (e.g. renaming types/fields)
+* Declaring which schemas "own" a particular type, and detecting conflicts
```
I think we should remove the use of the term “ownership” from the list of concerns as type ownership is a potential design pattern/anti-pattern depending on your point of view.

we can change this to a general call to conflict prevention/detection, which should go above the schema manipulation item directly above

n1ru4l replied:
I agree with this, something I miss in current implementations is field redundancy within multiple sub/schemas. E.g. multiple sub-schemas could implement a specific field and the gateway could then resolve the field on either of those sub-schemas. This is also something that would migrating a field from on sub-schema to another easier.

yaacovCR replied:
In terms of another concept related to ownership without endorsement of the pattern, we could add a final concern: “managing composition metadata with respect to handling the above concerns”

for instance, you may think Federation 1 type ownership is helpful because it obviates the need to annotate which schema’s type description is selected for shared types, versus @gmac ‘s introduction of canonical directive within schema stitching where the composition metadata is made explicit rather than flowing from the ownership model….

indeed, one of the immediate goals of this working group might be to consolidate around a set of directive names that should be reserved to represent composition metadata, which could happen well before the complete specification of the directives…

if there are other advantages of “ownership” not covered by the remaining categories that anyone can think of, I would suggest posting here and adding to the list

benjie replied:
Either way, "ownership" is a shared concern, whether the individual solutions decide that one schema owns the type versus shared ownership across all schemas.

yaacovCR replied:
Solutions without ownership have to carefully manage redundancy, which is a shared concern, but don’t have to concern themselves with ownership at all because shared ownership is kind of meaningless.

The term itself is nebulous and invites more questions than answers.

Not a huge deal, of course, but this is (imo) the solution space slipping into the description of the problem and should be avoided.

benjie replied:
Added based on feedback :+1:

mcollina commented on rfcs/CompositeSchemas.md:
```diff
@@ -0,0 +1,27 @@
+
+If this is of interest to you, please enter your name, GitHub handle, and organization below:
+
+| Name               | GitHub          | Organization       | Location
+| :----------------- | :-------------- | :----------------- | :-----------------
+| Benjie Gillam      | Benjie          | Graphile           | Chandler's Ford, UK
+| Michael Staib      | michaelstaib    | ChilliCream Inc.   | Zurich, CH
```
```suggestion
| Matteo Collina      | mcollina    | NearForm   | Forlì, IT
```

mcollina replied:
```suggestion
| Name               | GitHub          | Organization       | Location
| :----------------- | :-------------- | :----------------- | :-----------------
| Benjie Gillam      | Benjie          | Graphile           | Chandler's Ford, UK
| Michael Staib      | michaelstaib    | ChilliCream Inc.   | Zurich, CH
| Matteo Collina     | mcollina        | NearForm           | Forlì, IT
```

benjie commented:
Thanks for your interest, everyone! I've added everyone to the list that has attended a GraphQL Spec Working Group somewhat recently (since we added GitHub usernames to the agendas); for those that I could not add like this (@hayes @jensneuse @grillorafael @obi1kenobi @patrick91 @ndejaco2 @notrab @charlypoly) once this PR gets merged, please follow up with your own PR to add yourself which will require you to agree to the various GraphQL agreements (see: https://github.com/graphql/graphql-wg#contributing-to-this-repo), then we can get you added!

grillorafael commented:
Thank you very much @benjie. I will do that as soon as this PR gets in.

LunaticMuch commented:
I am all in, it I can still apply :)

doomsbuster commented:
I am interested!

awsed commented:
Count me in 🙌

andimarek commented:
@benjie 
Another relevant solution in this space is Nadel: https://github.com/atlassian-labs/nadel

benjie commented:
@andimarek Are you or someone from nadel interested in joining?

gmac commented:
I’m game.

OKJulian commented:
Awesome initiative, would love to contribute ❤️

linux-jedi commented:
Interested.

chikit commented:
I'm an engineer on the GraphQL server at Meta. Would like to contribute.

ataylormays-duetto commented:
I'm in! ✋

bhupatikrish commented:
Thank you for doing this. I would like to contribute, please count me in.

fotoetienne commented:
> a smaller specification that describes the common ground that everyone can build and innovate on top of

I think there would be a lot of value in this. In particular having a standard way to define object instance identity (such as @key in Apollo federation, or Node.id in Relay) would go a long way. Thanks for kicking this off @benjie !

benjie commented:
Hi folks! This sub-WG has been ratified, so now is the time to open a pull request to add yourself to the list of interested parties (if you're not already there): 

https://github.com/graphql/graphql-wg/edit/main/rfcs/CompositeSchemas.md

Next week is busy with the GraphQL conference, but after that I will look at the stated locations of all involved and try and pick an initial meeting time with the hopes as many of you as possible can make it. I will give at least a week's notice by posting in this thread.

Let's do this! 🙌 🚀

benjie commented:
First meeting is now scheduled; please see this issue for details: https://github.com/graphql/composite-schemas-wg/issues/1
